### DISCUSSION TITLE
How to use setValue with Controller input?

### DISCUSSION THREAD

**Discussion by cjambrosi** [react-hook-form/react-hook-form#9554] [Q&A] [answered]
In a flow. i want the form to come with some fields already filled in. I can't using with Controller.
Example:

```
const [currentEmail, setCurrentEmail] = useState('test@teste.com');

  const {
    handleSubmit,
    control,
    watch,
    register,
    formState: { errors, isValid, isDirty },
  } = useForm<IRegisterSubmitData>({
    resolver: yupResolver(schema),
    shouldUnregister: false,
    reValidateMode: 'onChange',
    mode: 'onChange',
    defaultValues: {
      email: '',
    },
  });

  useEffect(() => {
    if (currentEmail) setValue('email', currentEmail);
  }, [currentEmail]);


<Controller
  control={control}
  name="email"
  render={({ field: { onChange } }) => (
     <Input
         type="text"
         placeholder="Email"
         name="email"
         data-testid="email-field"
         onChange={onChange}
     />
  )}
/>
```

**Comment by bluebill1049** [comment]
you should be able to `setValue` against a controller-wrapped component, can you share a codsandbox where the issue is?

**Comment by Moshyfawn** [comment]
Sounds like a use-case for [the `reset` form method](https://react-hook-form.com/api/useform/reset), tbh.

```js
const { reset } = useForm()

useEffect(() => {
  if (data) {
    reset(data)
  }
}, [reset, reset])
```

**Comment by Moshyfawn** [comment, marked as answer]
One thing to note about your input code-snippet is that you don't pass `field.value` from the `render` method of `Controller`. Without passing the value input your controlled field, RHF can't set any values.

```diff
<Controller
  control={control}
  name="email"
- render={({ field: { onChange } }) => (
+ render={({ field: { value, onChange } }) => (
     <Input
         type="text"
         placeholder="Email"
         name="email"
         data-testid="email-field"
+        value={value}
         onChange={onChange}
     />
  )}
/>
```

**Reply by cjambrosi** [reply]
@Moshyfawn looking at the documentation, this is exactly what I needed 😄. Thank you!

**Reply by javierguzman** [reply]
Hello! I am facing the same issue for my form that uploads an avatar. The diference with this reply is that I do not have a component with an Input, it is just buttons + third part stuff from Expo (React Native):

```
<Controller
          control={control}
          render={({ field: { onBlur, onChange, value } }) => (
            <SelectImage
              onChooseImage={async (uri: string, ext: string) => {
                const { data, error } = await uploadImage('avatars', uri, ext);
                if (error) {
                  setCanFinish(false);
                  setError('avatar', {
                    type: 'custom',
                    message: 'Error uploading avatar'
                  });
                } else {
                  setCanFinish(true);
                  setValue('avatar', data.path);
                  console.log('Setting avatar');
                }
              }}
            />
          )}
          name="avatar"
        />
```

SelectImage will call onChooseImage function when the user selects an image from his mobile phone:
```
return (
    <View style={styles.container}>
      <View style={styles.previewSection}>{renderImageOrError()}</View>
      <View style={styles.selectionSection}>
        <Button
          buttonText={selectedImage ? 'Select another image' : 'Select image'}
          onPress={pickImageAsync}
          style={styles.selectButton}
        />
        <LoadingSpinner isLoading={isLoading} height={16} width={16} />
        {selectedImage && !isLoading && !usedImage && (
          <Button
            buttonText="Use this photo"
            disabled={isLoading}
            onPress={async () => {
              if (selectedImage) {
                setIsLoading(true);
                await props.onChooseImage(selectedImage, 'jpg');
                setIsLoading(false);
                setUsedImage(true);
              }
            }}
          />
        )}
      </View>
    </View>
  );
```

Any clue how to achieve the same here? Thank you in advance and regards

**Reply by Moshyfawn** [reply]
> Hello! I am facing the same issue for my form that uploads an avatar. The diference with this reply is that I do not have a component with an Input, it is just buttons + third part stuff from Expo (React Native):
> 
> ```
> <Controller
>           control={control}
>           render={({ field: { onBlur, onChange, value } }) => (
>             <SelectImage
>               onChooseImage={async (uri: string, ext: string) => {
>                 const { data, error } = await uploadImage('avatars', uri, ext);
>                 if (error) {
>                   setCanFinish(false);
>                   setError('avatar', {
>                     type: 'custom',
>                     message: 'Error uploading avatar'
>                   });
>                 } else {
>                   setCanFinish(true);
>                   setValue('avatar', data.path);
>                   console.log('Setting avatar');
>                 }
>               }}
>             />
>           )}
>           name="avatar"
>         />
> ```
> 
> SelectImage will call onChooseImage function when the user selects an image from his mobile phone:
> 
> ```
> return (
>     <View style={styles.container}>
>       <View style={styles.previewSection}>{renderImageOrError()}</View>
>       <View style={styles.selectionSection}>
>         <Button
>           buttonText={selectedImage ? 'Select another image' : 'Select image'}
>           onPress={pickImageAsync}
>           style={styles.selectButton}
>         />
>         <LoadingSpinner isLoading={isLoading} height={16} width={16} />
>         {selectedImage && !isLoading && !usedImage && (
>           <Button
>             buttonText="Use this photo"
>             disabled={isLoading}
>             onPress={async () => {
>               if (selectedImage) {
>                 setIsLoading(true);
>                 await props.onChooseImage(selectedImage, 'jpg');
>                 setIsLoading(false);
>                 setUsedImage(true);
>               }
>             }}
>           />
>         )}
>       </View>
>     </View>
>   );
> ```
> 
> Any clue how to achieve the same here? Thank you in advance and regards

Ideally, it's an input of type `"file"`. But you can easily connect a hidden `input` as it's a very valid approach as well

**Reply by javierguzman** [reply]
Thank you @Moshyfawn at the end it was a stupid error in the submit function I had, I leave it here just in case someone makes the same mistake:
```
const onSubmit = async (data) => {
    setIsLoading(true);
    setSignupData({
      ...signupData,
      ...data
    });
    const {
     otherstuffblablabla,
      avatar
    } = signupData;
```
So for sure "signup data" is not available until next rerender as I use setState(setSignupData) so inside onSubmit I should use "data" directly and not use signupData which uses setState. So it should be:
```
const {otherstuffFields} = signupData;
const { avatar } = data;
```
